### PR TITLE
[IR] LangRef: state explicitly that floats generally behave according to IEEE-754

### LINKED ISSUE #60942: Basic floating-point operations are underspecified
Currently, the LangRef does not specify the results of basic floating-point operations (`fadd`, `fsub`, `fmul`, `fdiv`) in any detail. APFloat uses IEEE 754 semantics, but the LangRef does not guarantee it. What guarantees are there about the behavior of floating-point code? If IEEE 754 is the intended model, then x87 codegen is completely broken and probably other targets are as well. If not IEEE 754, then what is the intended model?

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/llvm/llvm-project/issues/60942: IEEE semantics is likely what many frontends want (it definitely is what Rust wants), and it is what LLVM passes already assume when they use APFloat to propagate float operations.

This does not reflect what happens on x87, but what happens there is just plain unsound (https://github.com/llvm/llvm-project/issues/89885, https://github.com/llvm/llvm-project/issues/44218); there is no coherent specification that will describe this behavior correctly -- the backend in combination with standard LLVM passes is just fundamentally buggy in a hard-to-fix-way.

There's also the questions around flushing subnormals to zero, but [this discussion](https://discourse.llvm.org/t/questions-about-llvm-canonicalize/79378) seems to indicate a general stance of: this is specific non-standard hardware behavior, and generally needs LLVM to be told that basic float ops do not return the standard result. Just naively running LLVM-compiled code on hardware configured to flush subnormals will lead to #89885-like issues.

AFAIK this is also what Alive2 implements (@nunoplopes please correct me if I am wrong).

@nikic @arsenm @jcranmer-intel  what do you think is the best way to word this?

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -3577,6 +3577,12 @@ seq\_cst total orderings of other operations that are not marked
 Floating-Point Environment
 --------------------------
 
+Unless noted otherwise, LLVM works with IEEE 754 floating-point semantics: LLVM
+backends assume that the CPU is configured to provide IEEE-compatible behavior,
+and LLVM frontends can assume that LLVM IR floating-point operations behave
+according to the IEEE specification (with an :ref:`exception around NaN values
+<floatnan>`).
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:3584]: hyphenate IEEE-754? This is also worded a bit broadly, and seems to contradict some of the text later in the section.

Ideally the LangRef would not be in terms of "CPUs" or "configurations". 

Maybe should specify the bitlayouts are assumed to be the IEEE pattern?

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3584]: > This is also worded a bit broadly, and seems to contradict some of the text later in the section.

I assume you are talking about the exceptions related to NaN results? That's why it starts with "unless noted otherwise". I'm open for better wordings here if you have any suggestions.

> Maybe should specify the bitlayouts are assumed to be the IEEE pattern?

It already says that.

The goal of this PR is to say that `fadd` et al will use IEEE-754 rules to compute the desired results. That's a much stronger statement than just saying that the bitpatterns are interpreted as per IEEE-754.

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -3577,6 +3577,12 @@ seq\_cst total orderings of other operations that are not marked
 Floating-Point Environment
 --------------------------
 
+Unless noted otherwise, LLVM works with IEEE-754 floating-point semantics: LLVM
+backends assume that the CPU is configured to provide IEEE-compatible behavior,
+and LLVM frontends can assume that LLVM IR floating-point operations behave
+according to the IEEE specification (with an :ref:`exception around NaN values
+<floatnan>`).
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:3584]: Also the denormal exception

**Comment by jcranmer-intel** [comment]
I'm broadly supportive of the idea to bring greater clarity here; I'm not entirely certain this is the best way to do it. IEEE 754 just isn't quite precise enough to say that "it's just IEEE 754 behavior!" answers the questions that people are trying to ask. Also, we have three non-IEEE 754 types: `bfloat` (whose behavior is completely describable with IEEE 754 format parameters, although I think the denormal flushing behavior is likely to be especially inconsistent here), `x86_fp80` (which is mostly describable with IEEE 754, although it also introduces noncanonical values which don't exist with the basic IEEE 754 binary formats, but are allowed for in the standard), and `ppc_fp128` (which is very pointedly not an IEEE 754 format type, and I have yet to find a document describing its semantics in any useful detail).

What I think should be explicitly mentioned are these:
- it's not legal to x87-style store the result in a larger register without truncating it back to range (and you must ensure no double-rounding occurs when emulating in a larger type); `fadd float` really does mean IEEE 754 binary32 addition (as far as rounding is concerned).
- optimizations cannot change the precision of the result unless annotated with fast-math flags

(Denormal flushing is cursed for reasons beyond the scope of this PR, and I don't think the `denormal-fp-math` is really sufficient to capture the cursed nature.)

File: llvm/docs/LangRef.rst (hunk 2 above)
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3584]: What is the denormal exception? Is this about what happens when `denormal-fp-math` is set, but the default is to be IEEE-compatible?

**Comment by RalfJung** [comment]
> I'm not entirely certain this is the best way to do it. IEEE 754 just isn't quite precise enough to say that "it's just IEEE 754 behavior!" answers the questions that people are trying to ask.

It answers some questions. Like, currently the LangRef does not say what the result of `fadd` on  half, float, double, and fp128 even is. It says that the arguments and return values are interpreted using the IEEE-754 binary format, but doesn't say how the return value is computed -- so every frontend that relies on this being "the infinite-precision result, sounded to the next representable number, with round-ties-to-even" is technically incorrect. I think it should be possible to provide at least this basic promise.

I will restrict the new part to only apply to those types, not all float types.

> What I think should be explicitly mentioned are these:

To me these are both direct consequences of just saying unambiguously what the result is -- if the result can only be this, any optimization that changes the result is obviously wrong. But I can add clarifications to make sure this point is clear.

**Comment by RalfJung** [comment]
I have attempted to clarify these concerns, please take another look.

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -3572,6 +3572,29 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
 
+.. _floatsem:
+
+Floating-Point Semantics
+------------------------
+
+LLVM floating-point types fall into two categories:
+
+- half, float, double, and fp128, which correspond to the binary16, binary32,
+  binary64, and binary128 formats described in the IEEE-754 specification.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3583]: I think it would be useful to modify the Floating-point type section so that (e.g.) `half` is specifically described as "IEEE 754 binary16", in addition to the changes here.

File: llvm/docs/LangRef.rst [hunk 4]
```diff
@@ -3572,6 +3572,29 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
+For types that do correspond to an IEEE format, LLVM IR float operations behave
+like the corresponding operations in IEEE-754, with two exceptions: LLVM makes
+:ref:`specific assumptions about the state of the floating-point environment
+<floatenv>` and it implements :ref:`different rules for operations that return
+NaN values <floatnan>`.
+
+This means that optimizations and backends cannot change the precision of these
+operations (unless there are fast-math flags), and frontends can rely on these
+operations deterministically providing perfectly rounded results as described
+in the standard (except when a NaN is returned).
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3596]: I would add "This also means that backends are not allowed to implement floating-point instructions using larger floating-point types unless they take care to consistently narrow the results back to the original range without inducing double-rounding." or some similar text that makes it clear that mapping `fadd float` via just an x87 `FADD` instruction is not legal lowering.

File: llvm/docs/LangRef.rst [hunk 5]
```diff
@@ -3572,6 +3572,29 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
+- half, float, double, and fp128, which correspond to the binary16, binary32,
+  binary64, and binary128 formats described in the IEEE-754 specification.
+- The remaining types, which do not directly correspond to a standard IEEE
+  format.
+
+For types that do correspond to an IEEE format, LLVM IR float operations behave
+like the corresponding operations in IEEE-754, with two exceptions: LLVM makes
+:ref:`specific assumptions about the state of the floating-point environment
+<floatenv>` and it implements :ref:`different rules for operations that return
+NaN values <floatnan>`.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3591]: I'm struggling to come up with good wordsmithing here. It's not the case that there are two exceptions to IEEE-754 here. As @arsenm notes, we don't necessarily agree with IEEE 754 on denormal values. Also, we (rather explicitly) don't guarantee the side-effect on flags or exceptions of IEEE 754 semantics. The more we try to fine-tune the text to make "it's exactly IEEE 754 BUT", the more I think we obscure what the actual thing we're trying to state is.

This is a quick attempt I have of a direction that makes more sense to me:
> Unless otherwise specified, floating-point instructions return the same value [i.e., we make no implication about flags or exceptions here, although this may be subtle] that the corresponding IEEE 754 operation for their type when executing in the default floating-point environment [link], except that the behavior of NaN values is instead as specified here [link]. In particular, a floating-point instruction lacking fast-math flags operating on normal floating-point values is guaranteed to always return the same bit-identical result on all machines. When executing in the non-default floating-point environment, the result is typically undefined behavior unless constrained intrinsics [link] are used.

(This still doesn't cover denormal behavior, because that's non-default floating-point environment, but I'm not sure I have a firm grip on what the denormal guarantees we want to make in the first place are!)

File: llvm/docs/LangRef.rst (hunk 4 above)
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3596]: IMO that is covered by "backends cannot change the precision of these operations". If we start listing all the consequences of that statement, we'll never be done...

File: llvm/docs/LangRef.rst (hunk 3 above)
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3583]: Currently there is this sentence below that section:
"The binary format of half, float, double, and fp128 correspond to the IEEE-754-2008 specifications for binary16, binary32, binary64, and binary128 respectively."
So you are suggesting that should be moved inside the table? Sure, I can do that.

File: llvm/docs/LangRef.rst (hunk 5 above)
Comment (RalfJung) [reply llvm/docs/LangRef.rst]: I gave it a shot, based on your suggestion but structuring things a bit differently.

File: llvm/docs/LangRef.rst [hunk 6]
```diff
@@ -3572,6 +3572,44 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
+Various flags and attributes can alter the behavior of these operations and thus
+make them not bit-identical across machines and optimization levels any more:
+most notably, the :ref:`fast-math flags <fastmath>` as well as the ``strictfp``
+and ``denormal-fp-math`` attributes. See their corresponding documentation for
+details.
+
+If the compiled code is executed in a non-default floating-point environment
+(this includes non-standard behavior such as subnormal flushing), the result is
+typically undefined behavior unless attributes like ``strictfp`` and
+``denormal-fp-math`` or :ref:`constrained intrinsics <constrainedfp>` are used.
```
Comment (RalfJung) [inline llvm/docs/LangRef.rst:3611]: This paragraph is basically an exact duplicate of the second paragraph in the `floatenv` section, so I am inclined to remove it... but your draft did include such a sentence.

The way I view it, the `floatsem` section is just about the IEEE float formats. This paragraph is true for all formats so it should be in the `floatenv` section.

File: llvm/docs/LangRef.rst [hunk 7]
```diff
@@ -3572,6 +3572,39 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
+
+For floating-point operations acting on types with a corresponding IEEE format,
+unless otherwise specified the value returned by that operation matches that of
+the corresponding IEEE-754 operation executed in the :ref:`default
+floating-point environment <floatenv>`, except that the behavior of NaN results
+is instead :ref:`as specified here <floatnan>`. (This statement concerns only
+the returned *value*; we make no statement about status flags or
+traps/exceptions.) In particular, a floating-point instruction returning a
+non-NaN value is guaranteed to always return the same bit-identical result on
+all machines and optimization levels.
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3595]: Right now, LLVM assumes that all backends implement IEEE-754 arithmetic, and will miscompile code if the backend doesn't do that. One example of a target that does not implement IEEE-754 arithmetic is x86 without SSE, and https://github.com/llvm/llvm-project/issues/89885 has examples of code that gets micompiled due to that.

File: llvm/docs/LangRef.rst [hunk 8]
```diff
@@ -3572,6 +3572,39 @@ or ``syncscope("<target-scope>")`` *synchronizes with* and participates in the
...
+  format.
+
+For floating-point operations acting on types with a corresponding IEEE format,
+unless otherwise specified the value returned by that operation matches that of
+the corresponding IEEE-754 operation executed in the :ref:`default
+floating-point environment <floatenv>`, except that the behavior of NaN results
+is instead :ref:`as specified here <floatnan>`. (This statement concerns only
+the returned *value*; we make no statement about status flags or
+traps/exceptions.) In particular, a floating-point instruction returning a
+non-NaN value is guaranteed to always return the same bit-identical result on
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3594]: I meant all the operations that have an equivalent IEEE-754 operation. So yes that would include sqrt, though I was under the impression that it does not include transcendental functions.

I am not sure what is the best way to say that. Having a list seems awkward? Should each such operation have a comment, like "This corresponds to <op> in IEEE-754, so if the argument is an IEEE float format then the :ref:`floating-point semantics <floatsem>` guarantees apply."?

File: llvm/docs/LangRef.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:3595]: We're already at the point where we expect `float` et al to correspond to the IEEE 754 binary32 et al formats. (This is documented, although somewhat subtly, by the current LangRef). There is also agreement at this point that excess precision (à la x87) is not correct behavior for LLVM IR, although it's not (yet) explicitly documented in the LangRef.

The only hardware deviation from IEEE 754 that we're prepared to accept at this point is denormal handling. I'm reluctant to offer too many guarantees on denormal handling because I'm not up to speed on the diversity of common FP hardware with respect to denormals, but I'm pretty sure there is hardware in use that mandates denormal flushing (e.g., the AVX512-BF16 stuff is unconditionally default RM+DAZ+FTZ, with changing MXCSR having no effect).

In short, we already require that hardware supporting LLVM be IEEE 754-ish; this is tightening up the definition in the LangRef to cover what we already agree to be the case. In the putative future that we start talking about cases where `float` et al are truly non-IEEE 754 types (say, Alpha machines, or perhaps posits will make it big), then we can talk about how to add support for them in LLVM IR (which, given the history of LLVM, probably means "add new types", not "float means something different depending on target triple").

File: llvm/docs/LangRef.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:3594]: This is something that is hard to come up with a good term for. IEEE 754 has a core list of operations in section 5 which is a good starting point, but these omit the minimum/maximum operations (which are section 9.6). Section 9 is "recommended operations", and 9.2 is the main list of transcendental functions you're thinking of; IEEE 754 requires that they be correctly rounded, but C explicitly disclaims that requirement in Annex F. There's also a few functions in C that aren't in IEEE 754, notably ldexp and frexp.

(Note too that it was recently brought up in the Discourse forums that the libm intrinsics are meant to correspond to libm semantics, not IEEE 754 semantics.)

File: llvm/docs/LangRef.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3595]: > The only hardware deviation from IEEE 754 that we're prepared to accept at this point is denormal handling. 

Even there, the pass that causes trouble in https://github.com/llvm/llvm-project/issues/89885 would lead to miscompilations. `Analysis/ScalarEvolution` will assume that float ops that don't return NaNs produce a given bit pattern (including denormals), and if codegen later generates code that produces a different bit pattern, the result is a miscompilation.

So non-standard denormal handling is only supported with an explicit marker, which works very similar to the markers required for non-default FP exception handling.

File: llvm/docs/LangRef.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3594]: minimum/maximum don't do any rounding, making this clarification much less relevant. So maybe we should just say that this is about the core operations listed in section 5?

**Comment by RalfJung** [comment]
@jcranmer-intel @andykaylor I think I resolved/answered all of your comments, please let me know if there's anything else to do. :)

**Comment by RalfJung** [comment]
@nikic @arsenm maybe you could take a look? I am not sure who'd be a good reviewer for this. :)

File: llvm/docs/LangRef.rst [hunk 9]
```diff
@@ -3657,6 +3659,39 @@ specification on some architectures:
...
+This means that optimizations and backends may not change the observed bitwise
+result of these operations in any way (unless NaNs are returned), and frontends
+can rely on these operations providing perfectly rounded results as described in
+the standard.
+
+(Note that this is only about the value returned by these operations; see the
+:ref:`floating-point environment section <floatenv>` regarding flags and
+exceptions.)
+
+Various flags and attributes can alter the behavior of these operations and thus
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:3689]: and metadata (e.g. !fpmath)

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3689]: Ah, I didn't know about that one, thanks. I added a mention, and also used the opportunity to add links for `strictfp` and `denormal-fp-math`.

File: llvm/docs/LangRef.rst [hunk 10]
```diff
@@ -3657,6 +3661,40 @@ specification on some architectures:
...
+The value returned by those operations matches that of the corresponding
+IEEE-754 operation executed in the :ref:`default LLVM floating-point environment
+<floatenv>`, except that the behavior of NaN results is instead :ref:`as
+specified here <floatnan>`. In particular, such a floating-point instruction
+returning a non-NaN value is guaranteed to always return the same bit-identical
+result on all machines and optimization levels.
+
+This means that optimizations and backends may not change the observed bitwise
+result of these operations in any way (unless NaNs are returned), and frontends
+can rely on these operations providing perfectly rounded results as described in
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3684]: Ah, fair. It should be clear from context who is in charge of defining "correct" here (namely, IEEE-754).

I am adding these edits as new commits so it's easy to see what changed; I can squash them later or now if you prefer.

**Review by arsenm — APPROVED** [review]
I'm not 100% sure the correctly rounded bit applies to the default sqrt expansion on all targets, but I think we should consider that a bug if it's not.

**Comment by RalfJung** [comment]
@arsenm thanks for the review! I don't have merge rights, so may I ask you to handle that? :)
